core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/collection/LimitedSortedList.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright 2017 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.commonutil.collection;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import uk.gov.gchq.gaffer.commonutil.ToStringBuilder;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * <p>
+ * An {@link Iterable} which can sort, limit and deduplicate its elements.
+ * Sorting is achieved with a provided {@link Comparator}.
+ * </p>
+ * <p>
+ * This collection is backed by a {@link LinkedList},
+ * with items being dropped from the end of the backing collection.
+ * </p>
+ *
+ * @param <E> the type of object to store in the {@link LimitedSortedList}.
+ */
+public class LimitedSortedList<E> implements List<E> {
+    // Dummy value to associate with an Object in the backing Map
+    private static final Object PRESENT = new Object();
+
+    private final Comparator<E> comparator;
+    private final boolean deduplicate;
+    private final Integer limit;
+    private LinkedList<E> backingList;
+
+    // Used to deduplicate elements if required
+    private HashMap<E, Object> backingHashSet;
+
+    public LimitedSortedList(final Comparator<E> comparator) {
+        this(comparator, null);
+    }
+
+    public LimitedSortedList(final Comparator<E> comparator, final Integer limit) {
+        this(comparator, limit, false);
+    }
+
+    public LimitedSortedList(final Comparator<E> comparator, final Integer limit, final boolean deduplicate) {
+        this(comparator, limit, deduplicate, new LinkedList<>());
+    }
+
+    public LimitedSortedList(final Comparator<E> comparator, final Integer limit, final boolean deduplicate, final LinkedList<E> backingList) {
+        if (null == comparator) {
+            throw new IllegalArgumentException("Comparator is required");
+        }
+        if (null != limit && limit < 1) {
+            throw new IllegalArgumentException("Limit should be more than 0");
+        }
+        if (null == backingList) {
```

Comment:
Would we ever want to expose this constructor to allow users to specify their own backing list? Could we make this private?